irohad/ametsuchi/impl/redis_block_query.cpp

```diff
@@ -141,35 +150,96 @@ namespace iroha {
+                                 assets_id.end(),
+                                 [&transfer](auto const &a) {
+                                   return a == transfer.asset_id;
+                                 });
+        })
+             or isCommandValid<model::AddAssetQuantity>(
+        command, [&account_id, &assets_id](const auto &add) {
+          return add.account_id == account_id
+                 and std::any_of(assets_id.begin(),
+                                 assets_id.end(),
+                                 [&add](auto const &a) {
+                                   return a == add.asset_id;
+                                 });
+        });
+    }
 
-            for (auto block_id : block_ids) {
-              // create key for querying redis
-              std::string account_assets_key;
-              account_assets_key.append(account_id);
-              account_assets_key.append(":");
-              account_assets_key.append(std::to_string(block_id));
-              account_assets_key.append(":");
-              account_assets_key.append(asset_id);
-              client_.lrange(account_assets_key,
-                             0,
-                             -1,
-                             this->callbackToLrange(subscriber, block_id));
-            }
-            client_.sync_commit();
-            subscriber.on_completed();
-          });
+    rxcpp::observable<model::Transaction>
+    RedisBlockQuery::getAccountAssetTransactions(
+        const std::string &account_id,
+        const std::vector<std::string> &assets_id,
+        const model::Pager &pager) {
+      // TODO 06/11/17 motxx: Improve API for BlockQueries for on-demand
+      // fetching
+      return reverseObservable(
+        getBlocksFrom(1)
+          .flat_map([](auto block) {
+            return rxcpp::observable<>::iterate(block.transactions);
+          })
+          // local variables can be captured because this observable will be
+          // subscribed in this function.
+          .take_while([&pager](auto tx) {
+            return iroha::hash(tx) != pager.tx_hash;
+          })
+          .filter([this, &account_id, &assets_id](auto tx) {
+            return std::any_of(
+              tx.commands.begin(),
+              tx.commands.end(),
+              [this, &account_id, &assets_id](auto command) {
+                // This "this->" is required by gcc.
+                return this->hasAccountAssetRelatedCommand(
+                  account_id, assets_id, command);
+              });
+          })
+          // size of retrievable blocks and transactions should be
+          // restricted in stateless validation.
+          .take_last(pager.limit));
     }
 
+    // TODO 17/11/17 motxx - Discuss the effective solution with using Redis and Pagination
+    // approach with pagination.
+    /*
+        rxcpp::observable<model::Transaction>
+        RedisBlockQuery::getAccountAssetTransactions(const std::string
```

Comment:
It's better to paste both of this blocks in the pr for the discussion. It's hard to read w/o indention and syntax highlighting